Use the gray color for all abstract classes

MewPurPur opened the pull request:
This is meant as a small quality of life thing and something that should be "obviously better", so I haven't made a proposal for it yet. The changes in this PR are as follows:

- Gray out all abstract class icons that currently exist
- In Create Dialog and Search Help, use either NodeDisabled.svg for abstract nodes that don't have icons, or the newly added ObjectDisabled.svg for abstract non-node objects that don't have icons

I think this gives a pretty good result overall:

- You can more easily tell if a class is abstract from its icon

![image](https://github.com/godotengine/godot/assets/85438892/e9f17555-18b5-424c-ab68-a3590c126a7d)

- Virtual objects no longer appear like disabled nodes

![image](https://github.com/godotengine/godot/assets/85438892/30a0c2f2-cc49-436a-956d-66d502c2fa3a)

- Search Help now aligns with what you see in the creation dialog.

![image](https://github.com/godotengine/godot/assets/85438892/737b0153-1f7d-4879-9f52-54283b06babd)

(CollisionShape2D and CollisionShape3D currently appear with a white Object icon)

MewPurPur committed:
Use the gray color for all abstract classes

fire reviewed (approved):
As part of the usability team I really like the concept of greying out the images. Did not technical review.

Calinou reviewed (approved):
Tested locally, it works as expected.

![Screenshot_20230803_033615](https://github.com/godotengine/godot/assets/180032/dac5d40b-e8ea-41c1-aedd-0e30553affcf)

![Screenshot_20230803_033704](https://github.com/godotengine/godot/assets/180032/49d6c0b3-d4f4-4570-a16e-782d5c7f1eb2)

MewPurPur commented:
Missed something.

MewPurPur commented:
![image](https://github.com/godotengine/godot/assets/85438892/3b0cbe56-27e5-4f11-8668-23d9ac5f0f11)

Applied this also to the inspector.

![image](https://github.com/godotengine/godot/assets/85438892/d25e5032-3ce7-4171-a0f7-f1308b3bf368)

^ Light theme in case you're worried about contrast.

YuriSizov commented:
To add to that, we may sneakily use some of these icons for other visual elements where it can be important to preserve the color to convey the context/meaning.

MewPurPur commented:
This was actually my initial idea. I'm not against this per-se, but to be devil's advocate:

- Counterintuitively, doing this programmatically might lead to worse results than expected. The main thing is that, for example, to turn the light color into a disabled color, if our programmatic rule is to keep half of the original color, then we have to modulate not with `#808080`, but rather a really dark color `#1f1f1f`. I've not tested this though.
- It increases the amount of colors significantly, and with resources, these colors will often be used exactly once. Seeing an unfamiliar color might become unclear if it's something abstract or something you don't know. The color availability is reduced. It increases the risk of things blending in backgrounds.
- Less association between the 20% of abstract classes that have icons, and the 80% that are missing icons and so are using a grayed out node/object icon.

So I'll resist it a little for now, so we can have some discussion about these pros and cons.

@YuriSizov I don't get what you mean.

YuriSizov commented:
> I don't get what you mean.

These icons may be used for things other than node types in the editor. And in those cases it is very likely important that icons have their assigned color palette, as Akien mentions.

TheYellowArchitect commented:
>
    Doing this programmatically might lead to counterintuitive results. The main thing is that, for example, to turn the light color into a disabled color, if our programmatic rule is to keep half of the original color, then we have to modulate not with #808080, but rather a really dark color #1f1f1f. I've not tested this though.
    It increases the amount of colors significantly, and with resources, these colors will often be used exactly once. Seeing an unfamiliar color might become unclear if it's something abstract or something you don't know. The color availability is reduced. It increases the risk of things blending in backgrounds.
    Less association between the 20% of abstract classes that have icons, and the 80% that are missing icons and so are using a grayed out node/object icon.

Why have them programmatically? It's basically 3 colors. There is red (3D), green(Control), blue(2D) basically RGB. If the color isn't at least a shade of the aforementioned 3 colors, as Akien mentioned above, gray degrades the colorcoding intuitiveness.

MewPurPur commented:
Idea: 3 non-programmatical 60% gray colors for GUI, 2D, and 3D. Gray for white icons and everything else

akien-mga commented:
I'm pretty sure we can find a modulate value that makes it look desaturated without having to change the icons.

MewPurPur commented:
If we have a custom color for abstract 2D, 3D, and GUI nodes, then I'd prefer to have it in the icons so their appearance is consistent everywhere you find. IMO the only good point was about their color communicating something about their inheritance, and this would solve it.

MewPurPur commented:
![image](https://github.com/godotengine/godot/assets/85438892/ed50efb7-3304-4537-a333-2c4039913d57)

![image](https://github.com/godotengine/godot/assets/85438892/686e83ff-7844-450c-9d74-6f5dcef1c76d)

Zireael07 commented:
Related: #794 

We shouldn't be using solely color to convey inheritance

MewPurPur commented:
I saw that proposal, but my personal conclusion was as follows:
- The problem was always hypothetical, not a report from a colorblind person. In fact, no type of color blindness other than monochromia (extremely rare) should struggle to tell apart red and blue.
- White and green are problematic, but that's only a problem with Node, Node2D, Node3D, Control.

So then we're left with 10-15 nodes that share the same icon between blue and red. It seems like an obvious conclusion that a different shape improves usability, but I beg to differ:
- As aforementioned, colorblind users don't actually struggle to tell apart blue and red.
- Our icons are meant to work in 16x16. Flat icons are very easy to read and understand, but all suggestions that people updooted were for fancy 3D effects would be difficult to read in practice.
- The suggestion that didn't involve 3D effects had arbitrary pillars, like blue=masculine, red=feminine. People didn't like that.

Because of this, I only appreciate a change of 3D icons when there's an important functional difference it tries to communicate (e.g. CollisionPolygon3D).

NavigationObstacle and NavigationAgent use different shapes for their 2D and 3D versions' icons, even though they don't need to, and I think this shows the shortcomings of this whole idea. NavigationAgent3D is painful to recognize at small scales because it has this fancy 3D effect, whereas the simple flat NavigationAgent2D is insta recognizable.

So yes, I think improving this for monochromia people makes it way too much worse for everyone else for it to be worth it IMO.

And besides, just because colorblind people exist, doesn't mean we shouldn't use colors for usability. It rather means we shouldn't _rely_ on them and consider them more of a last resort, when nothing else works. And I don't think anything else works here.

YuriSizov commented:
The dimmed colors version looks nice, assuming it's done with modulation.

MewPurPur commented:
No, it's not modulation, it's three new color in the editor themes that are 60% gray and 40% green/red/blue.

Zireael07 commented:
> It rather means we shouldn't rely on them and consider them more of a last resort,

What I said, just in different words ;)

MewPurPur commented:
Not really, because I'm saying essentially "we should use color to represent inheritance because there's nothing better"

MewPurPur commented:
Next take: The colors used are a bit different, in a way that should be distinct enough from both plain gray and the node color. No modulation is used, I didn't get any nice colors from applying modulation indiscriminately.

I changed EditorNode's `get_class_icon()` function to get at the core of the issue, as there were more places listing nodes than just the 3 I addressed. Afaik now it looks good everywhere. The new `get_class_icon()` uses my logic as a default fallback for missing icons, and provides a way to use a different type of fallback - only used by EditorHelp, afaik.

aaronfranke commented on editor/create_dialog.cpp:
```diff
@@ -289,14 +289,12 @@ void CreateDialog::_configure_search_option_item(TreeItem *r_item, const String
 	bool can_instantiate = (p_type_category == TypeCategory::CPP_TYPE && ClassDB::can_instantiate(p_type)) ||
 			p_type_category == TypeCategory::OTHER_TYPE;
-	bool is_virtual = ClassDB::class_exists(p_type) && ClassDB::is_virtual(p_type);
+	bool instantiable = !(ClassDB::class_exists(p_type) && ClassDB::is_virtual(p_type)) && can_instantiate;
```
```suggestion
	bool instantiable = can_instantiate && !(ClassDB::class_exists(p_type) && ClassDB::is_virtual(p_type));
```

We should evaluate `can_instantiate` first, because, well, it's already evaluated.

aaronfranke reviewed (commented):
I like the colors in the latest version of the PR. I do wonder if it should be slightly grayer, but now it is easy to tell what dimension these nodes are from.

aaronfranke commented:
I just remembered I had already done something similar with the colors in [Node1D](https://github.com/aaronfranke/godot-1d):

<img src="https://raw.githubusercontent.com/aaronfranke/godot-1d/master/screenshots/1d_nodes.png" width="350" />

I used faded cyan for CollisionObject1D. So, yeah, I like the colors in this PR.

MewPurPur commented:
I opted for some desaturation because while cyan is a bright color, our 2D blue is even darker than the 50% gray.

Calinou commented:
How it looks with the latest revision of this PR:

Before | After
-|-
![Screenshot_20230830_203840](https://github.com/godotengine/godot/assets/180032/1d2a41b2-93b6-42c6-985f-15d03bef5228) | ![Screenshot_20230830_203824](https://github.com/godotengine/godot/assets/180032/629da5b5-7592-4792-a08f-4534fb8a27f3)

PS: Scrollbar, Slider and Separator abstract classes are lacking icons. Maybe we can use diagonally rotated variants like we did for BoxContainer?

YuriSizov commented on editor/editor_node.h:
```diff
@@ -868,7 +868,7 @@ class EditorNode : public Node {
 	StringName get_object_custom_type_name(const Object *p_object) const;
 	Ref<Texture2D> get_object_icon(const Object *p_object, const String &p_fallback = "Object");
-	Ref<Texture2D> get_class_icon(const String &p_class, const String &p_fallback = "Object");
+	Ref<Texture2D> get_class_icon(const String &p_class, bool p_default_fallback = true, const String &p_fallback = "ArrowRight");
```
Is this intentional? This shouldn't be changed this way, if so. If some tree/dialog needs this to be the default, then that's what it should pass to the method.

MewPurPur replied:
So changing the default from Object to ArrowRight shouldn't be done, is that the issue you're mentioning? I don't remember why I'd changed it, probaby

YuriSizov commented on editor/editor_help_search.cpp:
```diff
@@ -594,15 +594,11 @@ TreeItem *EditorHelpSearch::Runner::_create_class_hierarchy(const ClassMatch &p_
 
 TreeItem *EditorHelpSearch::Runner::_create_class_item(TreeItem *p_parent, const DocData::ClassDoc *p_doc, bool p_gray) {
 	Ref<Texture2D> icon = empty_icon;
```
This is unused now.

MewPurPur replied:
Okie I'll clean it up in this PR.

YuriSizov commented on editor/editor_node.cpp:
```diff
@@ -4349,15 +4349,25 @@ Ref<Texture2D> EditorNode::get_object_icon(const Object *p_object, const String
 	}
 
-	return _get_class_or_script_icon(p_class, scr, p_fallback, true);
+	String fallback = p_fallback;
+	// Use a white node icon if the class derives from Node, and a white object icon otherwise. Use a gray version if the class is abstract.
+	if (p_default_fallback && ClassDB::class_exists(p_class)) {
+		if (ClassDB::is_parent_class(p_class, SNAME("Node"))) {
+			fallback = (ClassDB::is_virtual(p_class) || !ClassDB::can_instantiate(p_class)) ? "NodeDisabled" : "Node";
+		} else {
+			fallback = ClassDB::is_virtual(p_class) ? "ObjectDisabled" : "Object";
+		}
+	}
```
I would argue that a similar logic should be used in `get_object_icon`. And for that reason I would put it into `_get_class_or_script_icon` in some way.

Perhaps, we can get rid of the boolean flag that you've added and treat empty `p_fallback` as a request to use some default value. And `p_fallback = ""` could be the default value of the argument in both methods.

YuriSizov reviewed (changes requested):
See above.

MewPurPur commented:
@Calinou Was already in my plans, I just wanted to finally determine what'll happen with abstract node icons.

MewPurPur commented:
Pushed fixes to the above.

AThousandShips commented on editor/editor_node.cpp:
```diff
@@ -4330,9 +4330,19 @@ Ref<Texture2D> EditorNode::_get_class_or_script_icon(const String &p_class, cons
+
+		// If the fallback is empty or wasn't found, use the default fallback.
+		if (ClassDB::class_exists(p_class)) {
+			bool instantiable = ClassDB::is_virtual(p_class) || !ClassDB::can_instantiate(p_class);
```
This is confusingly named

MewPurPur replied:
Why and how should I rename it?

AThousandShips replied:
The boolean is true when the class *isn't* instantiable, I'd either invert the boolean and swap the values below, or rename it to `non_instantiable`

YuriSizov reviewed (approved):
The implementation looks great now!

I've come around to the idea of hardcoding desaturated colors, so I think we can go ahead with this.

akien-mga commented:
Thanks!

aaronfranke commented:
With this PR, descendants of types without explicit icons are missing icons. I think this is a bug. Before:

<img width="172" alt="Screenshot 2023-09-04 at 1 12 46 AM" src="https://github.com/godotengine/godot/assets/1646875/a300ebea-0daa-4fee-8f6b-86c88af92d9d">

After: 

<img width="172" alt="Screenshot 2023-09-04 at 1 14 39 AM" src="https://github.com/godotengine/godot/assets/1646875/26b8ea14-382b-4c07-abce-236bfb3a2ad9">

YuriSizov commented:
> With this PR, descendants of types without explicit icons are missing icons.

Not entirely this PRs fault, but it did expose something.
